### PR TITLE
Property tests package for package#self.valid_name?

### PR DESCRIPTION
**Pull request by Ana06**
Property tests package for package#self.valid_name? and refactor package#self.valid_name? :bowtie:

### CODE REVIEW

**Comment by Ana06** [comment]
[(3) Add property tests to OBS](https://trello.com/c/VVho8Rf2/191-3-add-property-tests-to-obs)

File: src/api/app/models/package.rb [hunk 1]
```diff
@@ -989,16 +989,11 @@ def developed_packages
...
-    # obsolete, just for backward compatibility
-    return true if name =~ /\A_patchinfo:\w[-+\w\.]*\z/
-    return false if name =~ %r{[ \/:\000-\037]}
-    if name =~ %r{^[_\.]} && !%w(_product _pattern _project _patchinfo).include?(name)
-      return false
-    end
-    name =~ /\A\w[-+\w\.]*\z/ ? true : false
+    return true if %w(_product _pattern _project _patchinfo).include?(name)
+    # _patchinfo: is obsolete, just for backward compatibility
+    name =~ /\A([a-zA-Z0-9]|(_product:|_patchinfo:)\w)[-+\w\.]*\z/ ? true : false
```
Comment (bgeuken) [inline src/api/app/models/package.rb]: That would disallow +- and . as package name, right? Which was allowed before, in line 1001.

Comment (Ana06) [inline src/api/app/models/package.rb]: they were not allowed before :confused:

Comment (bgeuken) [inline src/api/app/models/package.rb]: I could build one:-) -> https://build.opensuse.org/package/show/home:bgeuken/does-that+work.

Comment (bgeuken) [inline src/api/app/models/package.rb]: You might want to extend your tests;-)

Comment (Ana06) [inline src/api/app/models/package.rb]: I thought you meant only `-` and `+` (one character name). `does-that+work` still works and my tests are testing it already!

Comment (bgeuken) [inline src/api/app/models/package.rb]: Ah, right. Because of the grouping

File: src/api/spec/models/package_spec.rb [hunk 2]
```diff
@@ -199,4 +200,68 @@
...
+        property_of {
+          string = sized(1) { string(/[a-zA-Z0-9]/) } + sized(range(0, 199)) { string(/[-+\w\.]/) }
+          guard string != '0'
+          string
+        }.check { |string|
+          expect(Package.valid_name?(string)).to eq(0)
+        }
+      end
+
+      it "starts by '_product:'" do
```
Comment (bgeuken) [inline src/api/spec/models/package_spec.rb]: starts with

File: src/api/spec/models/package_spec.rb [hunk 3]
```diff
@@ -199,4 +200,68 @@
...
+        property_of {
+          string = '_product:' + sized(1) { string(/[a-zA-Z0-9]/) } + sized(range(0, 190)) { string(/[-+\w\.]/) }
+          guard string != '0'
+          string
+        }.check(3) { |string|
+          expect(Package.valid_name?(string)).to be(true)
+        }
+      end
+
+      it "starts by '_patchinfo:'" do
```
Comment (bgeuken) [inline src/api/spec/models/package_spec.rb]: here as well

**Comment by bgeuken** [comment]
LGTM, beside the by / with mistake.

Feel free to merge after fixing that.

File: src/api/app/models/package.rb [hunk 4]
```diff
@@ -989,16 +989,11 @@ def developed_packages
...
     return false if name == "0"
-    return true if name =~ /\A_product:\w[-+\w\.]*\z/
-    # obsolete, just for backward compatibility
-    return true if name =~ /\A_patchinfo:\w[-+\w\.]*\z/
-    return false if name =~ %r{[ \/:\000-\037]}
-    if name =~ %r{^[_\.]} && !%w(_product _pattern _project _patchinfo).include?(name)
-      return false
-    end
-    name =~ /\A\w[-+\w\.]*\z/ ? true : false
+    return true if %w(_product _pattern _project _patchinfo).include?(name)
```
Comment (adrianschroeter) [inline src/api/app/models/package.rb]: A _project:$SOMETHING or _pattern:$someting is not valid, but it would become valid now

**Comment by adrianschroeter** [comment]
still, the valid_name? refactoring makes it harder to compare with the backend verification. Are you sure that it still matches the 000-037 rule for example? We can maybe adapt backend as well, but this is really critical code which can break a lot. You can compare it with BSVerify.pm verify_packid method:

sub verify_packid {
  my $packid = $_[0];
  die("packid is empty\n") unless defined($packid) && $packid ne '';
  if ($packid =~ /(?<!^_product)(?<!^_patchinfo):./) {
    die("packid '$packid' is illegal\n") unless $packid =~ /^([^:]+):([^:]+)$/s;
    my ($p1, $p2) = ($1, $2);
    die("packid '$packid' is illegal\n") if $p1 eq '_project' || $p1 eq '_pattern';
    verify_packid($p1);
    verify_packid($p2);
    return;
  }
  $packid =~ s/^_product://s;
  $packid =~ s/^_patchinfo://s;
  die("packid '$packid' is illegal\n") if $packid =~ /[\/:\000-\037]/;
  die("packid '$packid' is illegal\n") if $packid =~ /^[_\.]/ && $packid ne '_product' && $packid ne '_pattern' && $packid ne '_project' && $packid ne '_patchinfo';
  die("projid '$packid' is illegal\n") unless $packid;
  die("packid '$packid' is too long\n") if length($packid) > 200;
}

File: src/api/app/models/package.rb (hunk 4 above)
Comment (Ana06) [inline src/api/app/models/package.rb]: No... That would only return true if the name is exactly `_project` `_pattern` `_product`  or `_patchinfo`

**Comment by Ana06** [comment]
@adrianschroeter
 > Are you sure that it still matches the 000-037 rule for example? 

That checks that there are only ASCI characters right? And in the regular expression there are only ASCI characters, then yes it matches the rule.

File: src/api/app/models/package.rb (hunk 4 above)
Comment (Ana06) [inline src/api/app/models/package.rb]: I can also add a test for this case

**Comment by adrianschroeter** [comment]
Yes, ASCII characters, it forbids all special chars (before space char basically).

We may risk the change since the refactored code is indeed shorter, but it will be harder to verify on next bugreport when the backend behaves different (or when we need to adapt it)... That was the reason why we kept this code in sync as much as possible in the past (and we had problems here, because ruby regexp are working sometimes different then perl ones, but we finally solved the issues after some attempts).

Just to explain why it is causing some bad feeling in stomach when touching this;)

**Comment by adrianschroeter** [comment]
We (Michael, Ana, me) discussed this and are fine to merge it.

**Comment by bgeuken** [comment]
cool

**Pull request merged**